chacha20_poly1305/src/chacha20.rs

```diff
@@ -82,20 +80,24 @@ impl U32x4 {
 
     #[inline(always)]
     fn rotate_elements_left<const N: u32>(self) -> Self {
-        let mut result = [0u32; 4];
-        (0..4).for_each(|i| {
-            result[i] = self.0[(i + N as usize) % 4];
-        });
-        U32x4(result)
+        debug_assert!(N <= 4, "Invalid rotation value N={N}, only supports rotations up to 4");
```

Comment:
FWIW I had a go at finding another way to do this without `debug_assert` but could not. @kixunil might know a trick?

Reply:
Is `debug_assert` frowned upon? I'll admit I haven't used it much before.

Reply:
Just based on the thought that if we can do it at compile time its better than doing it at run time.

Reply:
In some contexts you can define `const _: () = { assert!(N <= 4); }` but IIRC it's not possible here. There's also `const { }` block but that requires quite recent compiler version. Perhaps a more principled (but boilerplate-y) approach:

```rust
trait UpTo4<const N: usize> {}

impl UpTo4<0> for () {}
impl UpTo4<1> for () {}
impl UpTo4<2> for () {}
impl UpTo4<3> for () {}
impl UpTo4<4> for () {}

fn rotate_elements_left<const N: usize>(self) -> Self where (): UpTo4<N> { /* ... */ }
```

Reply:
FWIW, here is how the experimental core library implements it, quite a bit more complexity for the general case: https://github.com/rust-lang/portable-simd/blob/master/crates/core_simd/src/swizzle.rs#L219

Ideally this custom version is dropped for the core version at some point in the future.  In the mean time, I think just need to strike a balance with complexity. @Kixunil's approach here is good enough for me.

Reply:
Updated in 36d45bf360b4fff069203a86e3429831babe5e5d